### PR TITLE
Qmake build hooks

### PR DESCRIPTION
**Pull request by abbradar**
###### Things done
- [ ] Tested using sandboxing (`nix-build --option build-use-chroot true` or [nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS)
- Built on platform(s)
  - [ ] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

This adds two hooks in spirit of `autoreconfHook`: `qmake4Hook` and `qt5{4,5}.qmakeHook`. Both are meant to automatically execute `qmake` on configure stage with necessary preparations and `PREFIX` argument. I've went through all packages that use Qt and half-automatically moved them to using those hooks. Also I did a little change in Qt 5 hook, moving final qmake setup to `postHook` from `preConfigureHook` -- this allows one to override `configurePhase` is he really wants to without additional headache.

cc @vcunat @ttuegel @aszlig (I've reverted your commits that add `preConfigure` in packages -- should perhaps have announced my to-be work earlier to avoid this).

WIP because this needs testing, although it builds several Qt 4/5 packages successfully. The problem is that I cannot test all of them right now -- several are broken, and `nox-review` doesn't work because of that.

### CODE REVIEW

**Comment by abbradar** [comment]
Oh, and previous discussion: https://github.com/NixOS/nixpkgs/issues/14670

**Comment by abbradar** [comment]
Unfortunately, it's difficult to filter generally broken packages from broken from this change. Nevertheless, I think this is stable enough.

**Comment by abbradar** [comment]
Ping? This should fix several packages and would bit-rot quickly if left laying.

File: pkgs/development/libraries/qt-5/5.5/qtbase/setup-hook.sh [hunk 1]
```diff
@@ -1,135 +1,3 @@
...
-    IFS="$delim" read -a dirs <<< "${!search}"
-    for dir in ${dirs[@]}; do
-        if [ "z$dir" == "z$target" ]; then exported=1; fi
-    done
-    if [ -z $exported ]; then
-        eval "export ${search}=\"${!search}${!search:+$delim}$target\""
-    fi
-}
-
-addToSearchPathOnce() {
```
Comment (ttuegel) [inline pkgs/development/libraries/qt-5/5.5/qtbase/setup-hook.sh]: @abbradar I need these functions to stay here or move to a setup hook that doesn't invoke qmake. They are used within the KDE packages, which all use cmake.

**Comment by ttuegel** [comment]
@abbradar Except for the issue I noted above, I think this is fine. I think rebuilding Qt should go into staging, though. I have a mass-Qt rebuild I'm about to merge into staging anyway.

File: pkgs/development/libraries/qt-5/5.5/qtbase/setup-hook.sh (hunk 1 above)
Comment (abbradar) [inline pkgs/development/libraries/qt-5/5.5/qtbase/setup-hook.sh]: No problem -- which ones do you need? All of them? I can look by myself if you point me at where to start looking.

Comment (ttuegel) [inline pkgs/development/libraries/qt-5/5.5/qtbase/setup-hook.sh]: @abbradar I just need the search path ones: `addToSearchPathOnceWithCustomDelimiter`, `addToSearchPathOnce`, and `propagateOnce`. Thanks!

Comment (abbradar) [inline pkgs/development/libraries/qt-5/5.5/qtbase/setup-hook.sh]: I'm building `kde5.kdelibs` to test if it works -- would it suffice?

**Comment by vcunat** [comment]
I just staged a bunch of mass rebuild commits (including stdenv-changing ones), so now it's a good time to add some more before Hydra starts rebuilding everything.

**Comment by abbradar** [comment]
I think I would finish test build in ~20 minutes -- do I have this time?

File: pkgs/development/libraries/qt-5/5.5/qtbase/setup-hook.sh (hunk 1 above)
Comment (ttuegel) [inline pkgs/development/libraries/qt-5/5.5/qtbase/setup-hook.sh]: @abbradar Fastest check would be `kde5.kcoreaddons` :smiley:

**Comment by vcunat** [comment]
Yes, I believe you have a couple hours at least. I can manage Hydra evaluations... and it will take a long time even to get stdenvs (which you don't seem to touch).

**Comment by ttuegel** [comment]
@vcunat Thanks for the heads-up! I just merged my pending Qt changes, so I'm ready to merge this into staging as soon as @abbradar is ready.

**Comment by vcunat** [comment]
:-) Merging such a large amount of changes at once will likely break something, but I think we can manage...

**Comment by ttuegel** [comment]
Well, KDE 5 doesn't start in master. I can assure you, it's at least less broken in that regard!

**Comment by abbradar** [comment]
Tested and applied few more fixes to several. packages. @ttuegel, please review https://github.com/NixOS/nixpkgs/pull/14770/commits/b857459bdbce58c833845b6adfdc68b1a204bed5 -- I've resolved merge conflicts with staging and maybe left one mistake or another.

**Comment by abbradar** [comment]
(Notice that this is based atop staging now -- don't hit "Merge"!)

**Comment by abbradar** [comment]
Pushed one more trivial fix.

**Comment by vcunat** [comment]
> Well, KDE 5 doesn't start in master.

Then there's nothing to fear :-)

**Comment by ttuegel** [comment]
@abbradar Looks good to me. All set on your end?

**Comment by abbradar** [comment]
Merged at https://github.com/NixOS/nixpkgs/commit/35cef635ffe8cfb75ff359dd5b68852b389b7ac2 -- thanks!

**Comment by vcunat** [comment]
Do you know about anything else to wait for before rebuilding?

**Comment by abbradar** [comment]
From my quick look through mass rebuild tagged PRs nothing enough agreed upon to merge right now.

**Comment by vcunat** [comment]
OK, building away: http://hydra.nixos.org/eval/1260112?compare=1260021

**Comment by vcunat** [comment]
Many kde5 packages fail due to having a cycle between the two outputs. (You can check the updated link above.)

**Comment by ttuegel** [comment]
Should be fixed in 9ea3e37. Part of the setup hook code for handling multiple outputs was moved to the qmake hook.

**Comment by abbradar** [comment]
Thanks for the fix, should have built more KDE packages to test...

**Comment by womfoo** [comment]
Hey @tokudan I've just prepared a fix for fritzing in #17043

**Pull request closed without merging**